### ISSUE TITLE
Notifications should be prioritised

### ISSUE THREAD

**Issue by zvirdaniel** [djdd87/SynoAI#145]
Hi! 

First of all, thanks for your work on this project, it's exactly what I was looking for and it's the reason I ended up with Synology NAS instead of dedicated NVR or some Raspberry setup. Config took some time, but it's working as expected.

However, I believe notificating the user should take priority before saving the image. Saving images takes time. In my case (slow drives and 4k camera) it takes up to 500ms. I suggest to send notification(s) before saving images.

![Uploading logs.png…]()

**Comment by djdd87** [comment]
_Maybe_... I think I wanted to do this but due to the way some of the notifications work the image had to be saved to a file first. I'll look into it again at some point.

**Comment by zvirdaniel** [comment]
I see. I suggest processing notificators which don't requires this first, then save the file, then process the rest. 

Would you be open to doing it this way? I have zero experience with .NET but I am experienced with some java development, so if you'd agree with doing it this way, I'd take a look.

**Comment by djdd87** [comment]
Yes, I'm definitely up for removing delays where possible. It looks like it's only Email that requires a FilePath, whereas everywhere else I return a FileStream from the saved image using File.OpenRead from GetReadOnlyStream on the ProcessedImage class.

It'd probably be doable by changing ProcessImage to take the SKBitmap object, then kicking off an async task to save the image so it's saved, e.g. Task<string> FilePath. Then Email can just use FilePath.Result.

The rest of the usages of GetReadOnlyStream could then return a new stream from SKBitmap.Encode.
